Fixed #24063 -- Locale code validation too strict

nijel opened the pull request:
The locale code can contain variant after @, so let's allow the
validation of it to pass.

berkerpeksag commented on tests/i18n/tests.py:
```diff
@@ -1424,8 +1424,11 @@ def test_check_for_language(self):
         self.assertTrue(check_for_language('en-us'))
         self.assertTrue(check_for_language('en-US'))
+        self.assertTrue(check_for_language('be'))
+        self.assertTrue(check_for_language('be@latin'))
```
Could you add more tests? For example, "tr_TR.UTF-8", "tr_TR.UTF8" (without the dash), "de_DE.utf-8" (lowercase encoding), "sr_RS@latin", "sr_RS@12345".

nijel replied:
Added, but specifying charset is currently not supported (and IMHO does not make sense as Django enforces UTF-8 on locales).

berkerpeksag commented:
Could you also add a release note to `docs/releases/1.8.txt` and squash the commits into one?

nijel commented:
Added release note and squashed commits as well.

berkerpeksag commented:
buildbot, test this please.

berkerpeksag commented on tests/i18n/tests.py:
```diff
@@ -1424,8 +1424,17 @@ def test_check_for_language(self):
+        self.assertTrue(check_for_language('be'))
+        self.assertTrue(check_for_language('be@latin'))
+        self.assertTrue(check_for_language('sr-RS@latin'))
+        self.assertTrue(check_for_language('sr-RS@12345'))
```
Looks like there is no standard for `@modifier`, but I'd reject all digits ones.  Or to be more strict, we can only accept iqtelif, euro, cyrillic, abegede, devanagari, latin: https://sourceware.org/git/?p=glibc.git;a=blob;f=localedata/SUPPORTED;h=9bf2a4f3840925c38bdefeaba28d2077835c7f07;hb=HEAD

nijel replied:
I've quite intentionally kept it matching anything (same as language or country codes), but we can indeed be more strict.

Anyway there can be more in use, for example ``ca@valencia``: https://github.com/translate/translate/blob/master/translate/lang/data.py#L44

nijel replied:
Anyway I don't think that hardcoding the list is good idea, here is list of variants available on my system:

```
sed -n '/@/ s/.*@\(.*\) .*/\1/p' /usr/share/i18n/SUPPORTED | sort -u
abegede
cyrillic
devanagari
euro
iqtelif
latin
saaho
valencia
```

collinanderson commented:
buildbot, test this please

timgraham commented on docs/releases/1.8.txt:
```diff
@@ -374,6 +374,10 @@ Internationalization
   reusable apps. It also allows overriding those custom formats in your main
   Django project.
 
+* Django now allows to use locale variants as supported by Gettext. These are
```
chop "as"
gettext (no caps)

timgraham commented on docs/releases/1.8.txt:
```diff
@@ -374,6 +374,10 @@ Internationalization
 
+* Django now allows to use locale variants as supported by Gettext. These are
+  usually used for languages which can be written in different scripts, for
+  example Latin and Cyrillic - eg. ``be@latin``.
```
, e.g.

timgraham commented:
Release note needs to be moved to 1.9 at this point.

nijel commented:
Okay, I've rebased changes on current master, moved the release notes and made suggested improvements in them.

Any chance to get this small change merged soon?

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -146,6 +146,10 @@ Internationalization
 * The :func:`django.views.i18n.javascript_catalog` view now works correctly
   if used multiple times with different configurations on the same page.
 
+* Django now allows to use locale variants supported by Gettext. These are
```
1. "Django" is a bit redundant give these are the Django release notes. How about "You can now use...". Also, can you add details about *where* in Django these locale variants can be used?
2. "gettext" shouldn't be capitalized

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -146,6 +146,10 @@ Internationalization
 
+* Django now allows to use locale variants supported by Gettext. These are
+  usually used for languages which can be written in different scripts, for
+  example Latin and Cyrillic - e.g. ``be@latin``.
```
dash -> comma

timgraham commented:
buildbot, add to whitelist.

timgraham commented:
merged in 76d26d89220f5740103a16bf71df2ae255c00da4, thanks!
